### ISSUE TITLE
MediaPlayerElement needs a better sample

### ISSUE THREAD

**Issue by Symbai** [windows-toolkit/Microsoft.Toolkit.Win32#106]
## I'm submitting a...
Sample app request 

## Current behavior
MediaPlayerElement has a short sample auto playing a web video

## Expected behavior
MediaPlayerElement needs also a sample demonstrating how to play / pause / stop / change volume and set / get playing position from code. Because there is no Play() method for example. And the only documentation I find is about the UWP version which seems to have a play method

https://docs.microsoft.com/en-us/uwp/api/Windows.UI.Xaml.Controls.MediaPlayerElement

which states:
`You can create your own media transport controls by setting AreTransportControlsEnabled to false, and using the Play and Pause methods on MediaPlayer. You can also control a rich set of properties by using the underlying MediaPlayer such as Position, Volume, IsMuted, IsLoopingEnabled, and PlaybackRate.`

## Environment

Nuget Package(s): 
Microsoft.Toolkit.Wpf.UI.Controls
Package Version(s): 
6.0.0-preview3

Windows 10 Build Number:
- [X] October 2018 Update (17763)

App min and target version:
- [X] October 2018 Update (17763)

Device form factor:
- [X] Desktop

Visual Studio 
- [X] 2019 (version: )

```

**Comment by michael-hawker** [comment]
@Symbai, the MediaPlayerElement is just a host, [as the documentation calls out for it](https://docs.microsoft.com/en-us/uwp/api/Windows.UI.Xaml.Controls.MediaPlayerElement#remarks), you need to access Play and other methods of it's [MediaPlayer](https://docs.microsoft.com/en-us/uwp/api/windows.media.playback.mediaplayer) property/reference.

**Comment by Symbai** [comment]
@michael-hawker Good idea except that MediaPlayer property throws a NullReferenceException all the time when being used in code. Which was the reason why I've asked for a sample because I thought I might be doing something wrong.

```
System.NullReferenceException: 'Object reference not set to an instance of an object.'

Microsoft.Toolkit.Wpf.UI.Controls.MediaPlayerElement.MediaPlayer.get returned null.
```

**Comment by michael-hawker** [comment]
@marb2000 @ocalvo @azchohfi any thoughts with your expertise in this area?  I thought the UWP MediaPlayerElement is supposed to have it's own MediaPlayer instance already initialized, so it should be retrievable here, eh?  Debugging the sample app seems to show this as the case as I see the MediaPlayer instance on the ChildInternal UWP object.

I thought maybe it'd be because the `Bind` method sets the default to one-way binding [here](https://github.com/windows-toolkit/Microsoft.Toolkit.Win32/blob/master/Microsoft.Toolkit.Wpf.UI.Controls/MediaPlayerElement/MediaPlayerElement.cs#L66).  But changing that doesn't seem to impact the property.

What do we need to change to get this to bubble up?

**Comment by michael-hawker** [comment]
@ocalvo Alex and I took a look it seems to be an initializing timing problem?  The `MediaPlayer` instance is `null` in `OnInitialized` in the wrapper, so the binding fails.  We tried setting up the binding in the Loaded event for the `UwpControl` itself where we see the MediaPlayer instance, but this also failed.  For some reason the WPF DependencyProperty is always returning null for MediaPlayer.

Any thoughts on what's happening here?  Thanks!

**Comment by ocalvo** [comment]
> @marb2000 @ocalvo @azchohfi any thoughts with your expertise in this area? I thought the UWP MediaPlayerElement is supposed to have it's own MediaPlayer instance already initialized, so it should be retrievable here, eh? Debugging the sample app seems to show this as the case as I see the MediaPlayer instance on the ChildInternal UWP object.
> 
> I thought maybe it'd be because the `Bind` method sets the default to one-way binding [here](https://github.com/windows-toolkit/Microsoft.Toolkit.Win32/blob/master/Microsoft.Toolkit.Wpf.UI.Controls/MediaPlayerElement/MediaPlayerElement.cs#L66). But changing that doesn't seem to impact the property.
> 
> What do we need to change to get this to bubble up?

My understanding is that it will only have its own instance when used from the WinUI Xaml compiler as only the Xaml compiler will initialize the media player instance (the WPF Xaml compiler does not count)

I think you will need to modify the wrapper to initialize the media player instance if Source property is set.

**Comment by Symbai** [comment]
Any update on this? I would really love to use this control but failing to access the mediaplayer property makes it impossible.

**Comment by michael-hawker** [comment]
We've pushed a fix for the MediaPlayer element being null.  It's now possible to access.  If anyone wants to improve the sample, please let us know.  Though we need to resolve #197 first.

**Comment by Symbai** [comment]
Finally, first time the mediaplayer becomes usable for offline purposes. Thank you so much ❤

**Comment by reinux** [comment]
MediaPlayer doesn't seem to have a Play() or Pause() method or anything like that. Am I looking at the wrong property?

**Comment by tonyhallett** [comment]
@reinux 
Another massive oversight.
The methods that you want are on this internal property and the wrapper class has not exposed them.
https://github.com/windows-toolkit/Microsoft.Toolkit.Win32/blob/1bc631b12a23b04acbcc38e7f682cb780d88b540/Microsoft.Toolkit.Win32.UI.Controls/Interop/WinRT/MediaPlayer.cs#L14

**Comment by tonyhallett** [comment]
A proposed solution for the null reference exception.

Set a dummy source ( 0 seconds duration ) instead of none.
Set the IsDummySource property to true.
When you change to a real source set IsDummySource to false. 

**This will not work currently as all wrapped controls are inherently broken both in terms of MVVM binding and changing properties in code.  I have fixed the Bind method of WindowsXamlHostBase which is at fault and will supply a pull request in the next couple of days.**

The code to add to MediaPlayerElement

```
       private bool hasLoaded;

        //add loaded handler to constructor
        protected MediaPlayerElement(string typeName)
            : base(typeName)
        {
            Loaded += MediaPlayerElement_Loaded;
        }

        private void MediaPlayerElement_Loaded(object sender, RoutedEventArgs e)
        {
            if (!hasLoaded)
            {
                EnableDisablePlayButton(!IsDummySource);
                hasLoaded = true;
            }
        }

        private void EnableDisablePlayButton(bool enabled)
        {
            MediaPlayer.SystemMediaTransportControls.IsPlayEnabled = enabled;
        }

        public bool IsDummySource
        {
            get { return (bool)GetValue(IsDummySourceProperty); }
            set { SetValue(IsDummySourceProperty, value); }
        }

        public static readonly DependencyProperty IsDummySourceProperty =
            DependencyProperty.Register("IsDummySource", typeof(bool), typeof(MediaPlayerElement), new PropertyMetadata(false, 
            (depObj, dp) =>
            {
                var isDummySource = (bool)dp.NewValue;
                var mediaPlayerElement = depObj as MediaPlayerElement;
                if (mediaPlayerElement.hasLoaded)
                {
                    mediaPlayerElement.EnableDisablePlayButton(!isDummySource);
                }
            }));
```

**Comment by Anoarai** [comment]
Hello,

I have ran into an issue where MediaPlayerElement instance returns null when MediaPlayerElement.MediaPlayer is called once again.

Am I missing something?

![image_2022-07-25_082053799](https://user-images.githubusercontent.com/97297612/180711172-9eb8ccfb-382d-4c13-8235-a7f1cd67d39b.png)
